swap/src/protocol/alice.rs

```
@@ -46,6 +50,162 @@ pub struct Swap {
+        swap_id: Uuid,
+        bitcoin_wallet: Arc<bitcoin::Wallet>,
+        monero_wallet: Arc<monero::Wallet>,
+        starting_balances: StartingBalances,
+        db_path: PathBuf,
+        listen_address: Multiaddr,
+    ) -> Self {
+        let network_seed = NetworkSeed::new(seed);
+        let identity = network_seed.derive_libp2p_identity();
+        let peer_id = PeerId::from(identity.public());
+
+        Self {
+            listen_address,
+            identity,
+            peer_id,
+            db_path,
+            swap_id,
+            bitcoin_wallet,
+            monero_wallet,
+            config,
+            starting_balances,
+        }
+    }
+
+    pub async fn new_swap_as_alice(&self, swap_amounts: SwapAmounts) -> Result<(Swap, EventLoop)> {
+        let initial_state = init_alice_state(
+            swap_amounts.btc,
+            swap_amounts.xmr,
+            self.bitcoin_wallet.clone(),
+            self.config,
+        )
+        .await?;
+
+        let (event_loop, event_loop_handle) = init_alice_event_loop(
+            self.listen_address.clone(),
+            self.identity.clone(),
+            self.peer_id.clone(),
+        )?;
+
+        let db = Database::open(self.db_path.as_path())?;
+
+        Ok((
+            Swap {
+                event_loop_handle,
+                bitcoin_wallet: self.bitcoin_wallet.clone(),
+                monero_wallet: self.monero_wallet.clone(),
+                config: self.config,
+                db,
+                state: initial_state,
+                swap_id: self.swap_id,
+            },
+            event_loop,
+        ))
+    }
+
+    pub async fn recover_alice_from_db(&self) -> Result<(Swap, EventLoop)> {
+        // reopen the existing database
+        let db = Database::open(self.db_path.clone().as_path())?;
+
+        let resume_state = if let database::Swap::Alice(state) = db.get_state(self.swap_id)? {
+            state.into()
+        } else {
+            unreachable!()
```

Why not return an error here?